include/zephyr/bluetooth/bluetooth.h

```diff
@@ -2005,6 +2021,26 @@ static inline int bt_le_whitelist_clear(void)
  */
 int bt_le_set_chan_map(uint8_t chan_map[5]);
 
+/**
+ * @brief Set the Resolvable Private Address timeout in runtime
+ *
+ * The new RPA timeout value will be used for the next RPA rotation
+ * and all subsequent rotations until another override is scheduled
+ * with this API.
+ *
+ * Initially, the if @kconfig{CONFIG_BT_RPA_TIMEOUT} is used as the
+ * RPA timeout.
+ *
+ * This symbol is linkable if @kconfig{CONFIG_BT_RPA_TIMEOUT_DYNAMIC}
+ * is enabled.
+ *
+ * @param new_rpa_timeout Resolvable Private Address timeout in seconds
+ *
+ * @retval 0 Success.
+ * @retval other RPA timeout value is invalid.
+ */
+int bt_le_set_rpa_timeout(uint16_t new_rpa_timeout);
```

Comment:
This API covers the following part of my rationale for this change:

> 4. Redefining RPA rotation periods in runtime (possibility to randomize RPA timeout improves privacy).

Yes, each advertising set has its own RPA value, but they all share a common RPA timeout which currently is configured statically using Kconfig (_CONFIG_BT_RPA_TIMEOUT_). One of the use cases for this API is to randomize RPA rotation periods. See the following code excerpt:

```
#define RPA_TIMEOUT_AVERAGE (15 * SEC_PER_MIN)
#define RPA_TIMEOUT_OFFSET_MAX  (1 * SEC_PER_MIN)

static uint16_t app_adv_rpa_timeout_randomize(void)
{
	int err;
	int8_t rand;
	uint16_t rpa_timeout = RPA_TIMEOUT_AVERAGE;

	err = sys_csrand_get(&rand, sizeof(rand));
	if (!err) {
		int16_t rand_timeout_diff;

		rand_timeout_diff = (((int)(RPA_TIMEOUT_OFFSET_MAX * MSEC_PER_SEC)) / INT8_MAX);
		rand_timeout_diff *= rand;
		rand_timeout_diff /= ((int) MSEC_PER_SEC);

		rpa_timeout += rand_timeout_diff;
	} else {
		LOG_WRN("Cannot get random RPA timeout (err: %d). Used fixed value", err);
	}

	return rpa_timeout;
}

static bool app_rpa_timed_out(struct bt_le_ext_adv *adv)
{
	int err;
	uint16_t next_rpa_timeout;

	LOG_INF("RPA timed out");

        /* Randomize the next RPA timeout. */
	next_rpa_timeout = app_adv_rpa_timeout_randomize();

	err = bt_le_set_rpa_timeout(next_rpa_timeout);
	if (err && (err != -EALREADY)) {
		LOG_ERR("bt_le_set_rpa_timeout failed: %d", err);
	} else {
		LOG_INF("Setting RPA timeout to %d [s]", next_rpa_timeout);
	}

        /* Update the advertising payload in sync with RPA timeout. */
	err = bt_le_ext_adv_set_data(adv, ad, ARRAY_SIZE(ad), sd, ARRAY_SIZE(sd));
	if (err) {
		LOG_ERR("Cannot set advertising data (err: %d)", err);
	}

	return true;
}
```

Reply:
If it is to add some randomness to the RPA timeout then it feels like a `CONFIG_BT_RPA_TIMEOUT_RANDOM_RANGE` could provide the same, i.e. a value range that is used to randomize the RPA timeout (within the range), without the owner of each advertising set needing to implement such a randomization function. 

With this approach it seems like `bt_le_set_rpa_timeout` will be called for each advertising set (all modifying the same global value), so even if one owner of an advertising set set the `bt_le_set_rpa_timeout` to e.g. 100, then it may be 10 or 1000 when the next advertising set has the `rpa_timed_out` called. 

I afraid I still don't see/understand the rationale behind changing the RPA timeout dynamically at runtime

Reply:
Yes, having the ```CONFIG_BT_RPA_TIMEOUT_RANDOM_RANGE``` Kconfig option as a default RPA timeout randomizer seems like a nice feature. However, some applications / protocols may have specific requirements on how this randomization process should look (specific algorithm). It seems beneficial to leave the user possibility of defining such a randomizer function.

The code example above demonstrates how the user can manage a single advertising set in their application. When there is more than one advertising set, it is up to the user how they want to randomize the RPA timeout across all advertising sets. They can choose one main advertising set, which drives the configuration of the RPA timeout for all advertising sets. Alternatively, they can have a dedicated function for setting the RPA timeout for all advertising sets and use it during the ```rpa_timed_out``` callback execution of one of the advertising sets. Ultimately, the user should be responsible for defining the RPA timeout when they use multiple advertising sets in their application.

I can also demonstrate one more use case for changing the RPA timeout in runtime. Imagine that we have protocols A and B with different RPA timeout requirements. When protocol A is active, protocol B is inactive (and the other way round). Switching between both protocols requires changing the RPA timeout.

And once again, the top-level rationale is to make it harder for an attacker to compromise the privacy of the device. Having a random RPA timeout makes it harder for an attacker to guess the RPA rotation. With this API, application developers can define the randomization algorithm for the RPA timeout to reflect their different levels of security requirements. Usually, more aggressive randomization translates to higher security within this single security aspect.

Reply:
Those are good points - Is there any existing profile or protocol right now that have those requirements? 

If so it would help if there's something tangible we can refer to as to why we need to introduce a new public API function. 

> However, some applications / protocols may have specific requirements on how this randomization process should look (specific algorithm). It seems beneficial to leave the user possibility of defining such a randomizer function.

One thing to consider, since the RPA timeout is global, any such requirements cannot be ensured with this API (e.g. if some profile/protocal requires an RPA timeout every minute (60 seconds) it would call this with `60`, but that is not stopping anything else from calling it with `600`, thus invalidating that requirement. Having a single static value (from a Kconfig) could at least indicate (and at buildtime) if two (or more) profiles/protocols are incompatible if they have different RPA timeout requirements (e.g. 60 and 600s).

Reply:
> Those are good points - Is there any existing profile or protocol right now that have those requirements?

Yes, there is the Fast Pair protocol from Google. You can find more details in the following link:

https://developers.google.com/nearby/fast-pair/specifications/configuration#BleAddress

> One thing to consider, since the RPA timeout is global, any such requirements cannot be ensured with this API (e.g. if some profile/protocal requires an RPA timeout every minute (60 seconds) it would call this with `60`, but that is not stopping anything else from calling it with `600`, thus invalidating that requirement. Having a single static value (from a Kconfig) could at least indicate (and at buildtime) if two (or more) profiles/protocols are incompatible if they have different RPA timeout requirements (e.g. 60 and 600s).

It is up to the application developer to somehow integrate conflicting requirements between protocols that they intend to use in the application.